Simplify retrieval of CMS objects to one cache request instead of two

Josniii opened the pull request:
One issue I've noticed with OctoberCMS is that it requires two cache requests to retrieve a CMS object like a page or a partial from the cache. First, the system calls `$cache->has($key)` to check if the key exists. This interally calls `get()`, resulting in a request being sent. Then, the system calls either `remember()` or `rememberForever()` to actually retrieve and/or store the result, resulting in another `get()` internally.

Since the first request is only used for a simple null check, I think it would be better to simply call `get()` and do the null check ourselves. Instead of calling `remember()` or `rememberForever()`, we call `put()`and `forever()`.

This way, we save a request when retrieving objects from the cache.

LukeTowers commented:
Doesn't retrieving an item from the cache leave it in memory?

Josniii commented:
I'd think that the cache server would keep it in memory, yes. However, the latency cost of sending a cache request can in some complex setups be pretty high.For example, if you're running an external cache server. In my testing with Elasticache, this resulted in a significant speed up in runtime.

LukeTowers commented:
Could this improvement be implemented on a lower level? I.e. implementing an in-memory cache for `Cache::get()` similar to the in-memory DB cache that we integrated a while ago? https://github.com/octobercms/library/pull/253

Josniii commented:
@LukeTowers - I looked into this a bit further and it looks like we already have [something like that.](https://github.com/octobercms/library/blob/master/src/Halcyon/MemoryCache.php)
At the end of `getCached()` the result is indeed stored in memory: 
```return MemoryCache::$cache[$key] = $result;```
And we do use any stored results as well:
```
if (array_key_exists($key, MemoryCache::$cache)) {
    return MemoryCache::$cache[$key];
}
```
So to me, it seems like we already have such a system implemented.

However, if we don't have the result in memory already, the cache retrieval is inefficient (because `has()` makes a request for the sole purpose of a null check), that's what this PR optimizes. To most people this is probably a minor issue, but when you have an external cache server, each request can take quite a while.

LukeTowers commented:
So the issue is that the cache driver being used doesn't implement an in-memory cache when `has()` is called?

Josniii commented:
I guess so. Ideally I'd say we cache the result that `has()` retrieves and we wouldn't have to worry about this. Any ideas on how to do that best?

LukeTowers commented:
That could be tricky, since it seems that the `has()` method is implemented by each cache driver separately; what's the repo for the cache driver that you're using? https://github.com/octobercms/library/blob/master/src/Halcyon/Builder.php#L703

Josniii commented:
`get_class($cache)` gives me `Illuminate\Cache\Repository` - so that would be the [Laravel repository.](https://github.com/laravel/framework). I'm using memcached.

LukeTowers commented:
Well, we could either try to make a PR to laravel adding in memory cache for get(), or we could put our own layer on top of it and use dependency injection to replace it with our layer that extends theirs; implementing the in memory cache in our layer. Basically the point of this entire repo which is to put that layer between OctoberCMS projects and Laravel.

@Josniii Thoughts?

Josniii commented:
@LukeTowers Not sure, but I think that in-memory cache might be a bit too general for Laravel since it is used for so many different types of applications? It makes sense in a CMS where we like to reuse small fragments a lot - maybe not so much for a lot of other types of applications. So I'd say the best place to put such a layer would be in October.

Josniii commented:
Here's how I think we could do this. If you could see a more elegant way of implementing this I'd like some input:

- Replace the default `CacheManager` with a `MemoryCacheManager` through dependency injection, that extends the default one. This could be done in the `HalcyonServiceProvider`. Override `repository()` so it instances a custom MemoryRepository class. We could put a config check here so people can decide if they want this enabled or not.
- Implement a `MemoryRepository` that extends the Laravel `Repository`. Here we can override `get()` so that it stores the result in memory before calling `parent::get()` to retrieve things the first time. We'll probably also need to override any functions used to write to the cache, that way we can update the cache as needed as well.

How does this sound?

Josniii commented:
Hi @LukeTowers 
I finally had some time to take a crack at this.
I've changed this around to try and implement what I suggested above. I might have missed some neccessary overrides for the `MemoryRepository`, but it seems to work fine. My testing sees a large decrease in external cache requests as hoped.

LukeTowers commented on src/Halcyon/Builder.php:
```diff
@@ -244,8 +244,8 @@ public function first()
      */
-     public function get($columns = ['*'])
-     {
+    public function get($columns = ['*'])
```
Indentation 4 spaces please

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,136 @@
+     */
+    public function flush()
+    {
+        $this->cache = [];
```
Should this be a call to `$this->flushInternalCache()` instead?

LukeTowers commented:
@Josniii awesome work, looking forwards to getting this merged! What are you ideas around making this behaviour configurable? Also, would it be possible to add some unit testing for this functionality?

Josniii commented:
@LukeTowers I think we should add a flag to ```config/cache.php``` like `'request_cache' => true,` to let people opt-out of the feature if they have something else that conflicts with this. Other than that, I don't really have any other suggestions. I've added ```MemoryRepositoryTest.php``` with tests for the feature, let me know if you think I'm missing anything 👍

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,166 @@
+     * @param null $default
+     * @return mixed
+     */
+    public function getFromMemoryCache($key)
```
If these methods are used in testing, then they need to be what the other (`get` & `put`) methods call to interact with the memory cache

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,166 @@
+     * Used in testing
+     *
+     * @param $key
+     * @param null $default
```
Remove this please

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,166 @@
+    public function forget($key)
+    {
+        unset($this->cache[$key]);
+        return parent::forget($key); // TODO: Change the autogenerated stub
```
Deal with TODO

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,166 @@
+    public function forever($key, $value)
+    {
+        $this->cache[$key] = $value;
+        parent::forever($key, $value); // TODO: Change the autogenerated stub
```
Deal with TODO

LukeTowers commented on tests/Halcyon/MemoryRepositoryTest.php:
```diff
@@ -0,0 +1,81 @@
+    {
+        $this->repository->put('foo', 'bar', 10);
+        $this->repository->forget('foo');
+        $this->assertNull($this->repository->get('foo'));
```
Should this be getFromMemoryCache instead?

LukeTowers commented on tests/Halcyon/MemoryRepositoryTest.php:
```diff
@@ -0,0 +1,81 @@
+        $this->repository->put('foo', 'bar', 10);
+        $this->repository->put('baz', 'qux', 10);
+        $this->repository->put('quux', 'corge', 10);
+        $this->assertEquals('bar', $this->repository->get('foo'));
```
Should these `get()` calls be changed to `getFromMemoryCache()` for testing the memory cache specifically?

LukeTowers commented:
Re the config, I think having `'disableRequestCache' => false` added to `config/cache.php` with an appropriate comment block would be ideal. This will enable the cache by default for updates (since updating October does not modify config files) while still providing the option to disable it if required. If you could submit a PR to octobercms/october; that would be awesome!

Thanks again for your work on this so far @Josniii, I appreciate it!

Josniii commented:
@LukeTowers I've gone through your reviews and made changes based on your suggestions, please take a look and let me know if you have further comments. I'll try and get the PR for the CMS repository done tomorrow, I'll tag you there once it's ready. Thanks for being thorough!

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,170 @@
+            return $this->many($key);
+        }
+
+        if($value = $this->getFromMemoryCache($key)) {
```
Does this support a falsey value being stored in the cache?

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -29,7 +29,9 @@ public function get($key, $default = null)
-        if($value = $this->getFromMemoryCache($key)) {
+        $value = $this->getFromMemoryCache($key);
+
+        if(isset($value)) {
```
I'd prefer that you use `!is_null($value)` instead for this, since that is what you are actually testing for, the fact that `isset()` returns `false` for `null` is not immediately obvious unless you're already aware of that fact so for readability I would prefer the `!is_null()` check instead. Thoughts?

Josniii replied:
I agree, I was actually checking out whether or not `isset()` returns true for `null` myself. Better to just use `!is_null()` in the first case.

Josniii commented:
I've added a PR to OctoberCMS [here.](https://github.com/octobercms/october/pull/3477) I've also implemented `isEnabled` in `MemoryCacheManager` so it hooks to this config flag.

LukeTowers commented on src/Halcyon/MemoryCacheManager.php:
```diff
@@ -0,0 +1,19 @@
+
+    public static function isEnabled()
+    {
+        return !Config::get('cache.disableRequestCache', false);
```
Are you able to use the `$this->app->config` methodology for accessing the config value instead of pulling in the Config facade? It's used in other places in this repo so I'd prefer to use it here too

Josniii replied:
I don't think I can since this is a static function, unless I'm missing something? I used the facade because I had seen that other places as well.

LukeTowers replied:
Fair enough, you'd have to implement it as a facade which would be a bit ridiculous

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,172 @@
+     * Retrieve an item from the cache by key.
+     *
+     * @param  string|array $key
+     * @return mixed
```
I know I said to remove the `@param $default` that was here earlier but I swear the `$default = null` parameter wasn't present in the method signature when I said that :) Could you re-add it?

Josniii replied:
I think that was in `getFromMemoryCache` I removed it. So it's just missing here :)

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,172 @@
+    public function put($key, $value, $minutes = null)
+    {
+        if (is_array($key)) {
+            $this->putMany($key, $value);
```
See https://github.com/laravel/framework/pull/13868#issuecomment-223764289 for a reason why the `is_array` check should be removed entirely

LukeTowers commented on src/Halcyon/MemoryRepository.php:
```diff
@@ -0,0 +1,172 @@
+    public function put($key, $value, $minutes = null)
+    {
+        if (is_array($key)) {
+            $this->putMany($key, $value);
```
See https://github.com/laravel/framework/issues/23729 for a reason why `is_array` check should be removed entirely here.

LukeTowers replied:
@taylorotwell has decided to not remove it (even though it's not documented anywhere and the argument mapping doesn't line up) so we'll have to leave it here for compatibility with Laravel.

Josniii replied:
Indeed, I agree it looks very strange, although I guess it technically works. You'd have to call it like `put(['foo' => 'bar'], 10)` with the `$value` param effectively being used as `$minutes`.
